be2-scala/src/test/scala/ch/epfl/pop/authentication/AuthenticateSuite.scala

```diff
@@ -0,0 +1,222 @@
+    val params = parametersNames.zip(
+      List(response_type, client_id, redirect_uri, scope, state, response_mode, login_hint, nonce)
+    )
+    buildRequest(params)
+  }
+
+  private def getAttributeValue(response: HttpResponse, key: String) = {
+    val attributes = response.attributes
+    val attributesWithName = attributes.map(pair => pair._1.name -> pair._2)
+    attributesWithName.get(key)
+  }
+
+  test("valid request succeeds") {
+    val route = Authenticate.buildRoute()
+    val request = buildRequest(
+      responseType,
+      clientID,
+      redirectUri,
+      scope,
+      state,
+      responseMode,
+      loginHint,
+      nonce
+    )
+
+    request ~> route ~> check {
+      status shouldBe OK
+    }
+  }
+
+  test("valid request without optional params succeeds") {
+    val route = Authenticate.buildRoute()
+    val request = buildRequest(
+      responseType,
+      clientID,
+      redirectUri,
+      scope,
+      "",
+      "",
+      loginHint,
+      nonce
+    )
+
+    println(request)
+
+    request ~> route ~> check {
+      status shouldBe OK
+    }
+  }
+
+  test("invalid response type fails request") {
+    val badResponseType = "invalid"
+
+    val route = Authenticate.buildRoute()
+    val request = buildRequest(
+      badResponseType,
+      clientID,
+      redirectUri,
+      scope,
+      state,
+      responseMode,
+      loginHint,
+      nonce
+    )
+
+    request ~> route ~> check {
+      status shouldBe Found
+      getAttributeValue(response, "error") shouldBe Some("unsupported_response_type")
+    }
+  }
+
+  test("invalid scope fails request") {
+    val badScope = "invalid x y"
+
+    val route = Authenticate.buildRoute()
+    val request = buildRequest(
+      responseType,
+      clientID,
+      redirectUri,
+      badScope,
+      state,
+      responseMode,
+      loginHint,
+      nonce
+    )
+
+    request ~> route ~> check {
+      status shouldBe Found
+      getAttributeValue(response, "error") shouldBe Some("invalid_scope")
+    }
+  }
+
+  test("missing parameter fails request") {
+    val route = Authenticate.buildRoute()
+
+    val mandatoryNames = parametersNames.filter(name => name != "state" && name != "response_mode")
+    val mandatoryValues = List(responseType, clientID, redirectUri, scope, loginHint, nonce)
+    val mandatoryParams = mandatoryNames.zip(mandatoryValues)
+
+    for (name <- mandatoryNames) {
+      val subParams = mandatoryParams.filter(_._1 != name)
+      val request = buildRequest(subParams)
+
+      request ~> route ~> check {
+        status shouldBe Found
+        getAttributeValue(response, "error") shouldBe Some("invalid_request")
+        getAttributeValue(response, "error_description") shouldBe Some(s"Missing parameters: [$name]")
+      }
+    }
+  }
```

Comment:
A lot seems to be going there, I am not sure to follow, some comments would help ;)

Reply:
I simplified the code a bit and changed some variables names to make it clearer